### ISSUE TITLE
files stuck in Publication NEW status

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6450]
following up on user report
https://hypernews.cern.ch/HyperNews/CMS/get/crabFeedback/8537.html
about task
https://cmsweb-k8s-prod.cern.ch/crabserver/ui/task/210227_061522%3Aaaravind_crab_20210227_001455

6 files are stuck in publication NEW status and somehow never picked up by Publisher.

This
https://cmsweb-k8s-prod.cern.ch:8443/crabserver/prod/workflow?subresource=publicationstatus&workflow=210227_061522:aaravind_crab_20210227_001455&username=aaravind&asodb=filetransfers&asourl=https://cmsweb.cern.ch/crabserver/prod
returns
```
{"result": [
 {"status": {"new": 6, "done": 3994}}
]}
```

and Publisher log simply has no mention of that tasl after March 2nd.

**Comment by belforte** [comment]
Publication status of individual files can be found at
https://cmsweb.cern.ch/crabserver/prod/fileusertransfers?subresource=getPublicationStatus&taskname=210227_061522:aaravind_crab_20210227_001455&username=aaravind

Publication Status codes:
https://github.com/dmwm/CRABServer/blob/afab4af9251d3ccfa33b3158750ef176a86cbe38/src/python/ServerUtilities.py#L93-L98

**Comment by belforte** [comment]
the 6 stuck files are:
```
{"desc": {"columns": ["tm_id", "tm_jobid", "tm_publication_state", "tm_start_time", "tm_last_update"]}, "result": [
,["be5e949ef2f1240547c1336de1a8fcf52ad703835d9cdc72aa30dc83", "339", 0, 1614462775, 1614462775]
,["9307d5d48615e072f2cee333d28aef270de7cd79a8e2f2e6184f74e8", "552", 0, 1614463313, 1614463314]
,["61d5ef45adcb28111a0e166ba34e41c76ca8d06cc990e3bb57bbaabd", "3471", 0, 1614458824, 1614458824]
,["8cf1744f88126e6fa743dbef80b771482e00ab8460ffb415a62bdd83", "1006", 0, 1614458367, 1614458368]
,["1c85d2c06f4c7565edbb773c41edba7eca0e38708d003e92d84a3efa", "188", 0, 1614459166, 1614459166]
,["37e89f910afb8ca1ef83ecd1a48978d4e431f0be72effe265077eef8", "668", 0, 1614449827, 1614449828]
```

**Comment by belforte** [comment]
looking at transfer status for those files via crabserver/ui, they are all in NEW
and therefore Publisher rightfully does not pick them.
 
![Screenshot from 2021-03-09 15-12-32](https://user-images.githubusercontent.com/1837785/110483315-e6a15480-80e9-11eb-8453-5f6694dcc64a.png)

**Comment by belforte** [comment]
looking at jobid 188. PostJob shows no errors:
https://cmsweb.cern.ch:8443/scheddmon/0120/cms1772/210303_171138:abala_crab_crab_L5JERC_EGamma_2018D_CP5_Autumn18_JECV19/postjob.188.0.txt

**Comment by belforte** [comment]
But PostJob get status of transfer from the file `"aso_status.json"`, not from transfersdb
https://github.com/dmwm/CRABServer/blob/afab4af9251d3ccfa33b3158750ef176a86cbe38/src/python/TaskWorker/Actions/PostJob.py#L973

**Comment by belforte** [comment]
looking at what FTS_transfers did for the first file
```
id: 1c85d2c06f4c7565edbb773c41edba7eca0e38708d003e92d84a3efa   job 188
```
I find
```
belforte@vocms0198/transfers> grep 1c85d2c06f4c7565edbb773c41edba7eca0e38708d003e92d84a3efa transfer_inject.log 
2021-02-27 14:38:29,215[ 10930]MainThread: Marked submitted ['a31f04e407bd77078d2ab8a7ddc2f1498ad9c8f3e24d50803754ab2d', '41c5c3858381d6e9f7931c5d8effdbe79d2afaedec5b513d7da67c6e', 'efcb1d465ae8f92bc36d44eae2d9036bf83834c31b14900f24423f2e', '1c85d2c06f4c7565edbb773c41edba7eca0e38708d003e92d84a3efa'] files
2021-02-27 14:43:38,054[  4975]MainThread: Marked failed [u'a31f04e407bd77078d2ab8a7ddc2f1498ad9c8f3e24d50803754ab2d', u'1c85d2c06f4c7565edbb773c41edba7eca0e38708d003e92d84a3efa']
belforte@vocms0198/transfers> 

```
things do not make sense. 

the lines surrounding the `Marked failed` are puzzling:
```
2021-02-27 14:43:38,054[  4975]MainThread: Marked failed [u'a31f04e407bd77078d2ab8a7ddc2f1498ad9c8f3e24d50803754ab2d', u'1c85d2c06f4c7565edbb773c41edba7eca0e38708d003e92d84a3efa']
2021-02-27 14:43:38,055[  4975]MainThread: Marking job 1 files done and 0 files failed for job 12ee2176-7901-11eb-a676-fa163e36d89b
```
the FTS jobid recorded in filetransfersdb is `122990f4-7901-11eb-895d-fa163e36d89b`

Looking more at the filetransfersdb record for this file:
```
tm_last_update | 1614459166
tm_start_time | 1614459166
```
i.e. the record was never updated since creation. Somehow FTS_transfers.py silently failed to update the status.

**Comment by belforte** [comment]
but  FTS_transfers has proper exception catchig, so it looks like POST call returned success but DB was not updated
https://github.com/dmwm/CRABServer/blob/afab4af9251d3ccfa33b3158750ef176a86cbe38/scripts/task_process/FTS_Transfers.py#L82-L107

**Comment by belforte** [comment]
In conclusion, this is a very obscure and rare problem. I do not thing that adding extra DB calls to sort of repeat the DB updates will make any sense and my only degrade performances.

Only sensible strategy is to improve crab status to list the ID of the stuck jobs so that the user can do a crab resubmit for those.

**Comment by belforte** [comment]
it happens too often. There must be something wrong in the code which could possibly be improved.

**Comment by belforte** [comment]
e.g. this
![Screenshot from 2021-03-12 18-29-34](https://user-images.githubusercontent.com/1837785/110976245-f7a7cb00-8360-11eb-8573-d7b560fc8446.png)
job was run again, all good on retry 1. but transfersdb was not updated and file will not be published.

One way could be to force again trasnfer status to done when PJ updates the publication status, but.. why didn't FTS_transer keep the DB up to date ??

**Comment by belforte** [comment]
in this last task of mine https://cmsweb.cern.ch/crabserver/ui/task/210312_145732%3Abelforte_crab_20210312_155727/dbver/preprod
FTS_transfer logs actually reports that the file was first marked as submitted and then as failed
```
belforte@vocms059/transfers> grep bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f transfer_inject.log 
2021-03-12 16:15:45,096[ 45781]MainThread: Marked submitted ['bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f'] files
2021-03-12 16:21:11,543[ 23313]MainThread: Marked failed [u'bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f']
belforte@vocms059/transfers> 
```
which correspond to the attempts during 1st execution of job #1159
```
belforte@vocms059/SPOOL_DIR> grep bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f postjob.1159.0.txt 
Fri, 12 Mar 2021 16:10:06 CET(+0000):INFO:PostJob LFN /store/temp/user/belforte.be1f4dc5be8664cbd145bf008f5399adf42b086f/GenericTTbar/Stefano-Test-210312/210312_145732/0001/kk_1159.root (id bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f) is not in ASO database. Will inject a new transfer and publication request.
Fri, 12 Mar 2021 16:10:06 CET(+0000):INFO:PostJob ASO job description: {'_id': 'bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f',
Fri, 12 Mar 2021 16:40:07 CET(+0000):ERROR:PostJob Stageout job (internal ID bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f) failed with status 'failed'.
Fri, 12 Mar 2021 16:40:07 CET(+0000):INFO:PostJob There were 1 failed/killed transfers: bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f
```
but then job was resubmitted and everything worked:
```
belforte@vocms059/SPOOL_DIR> grep bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f postjob.1159.0.txt 
Fri, 12 Mar 2021 16:10:06 CET(+0000):INFO:PostJob LFN /store/temp/user/belforte.be1f4dc5be8664cbd145bf008f5399adf42b086f/GenericTTbar/Stefano-Test-210312/210312_145732/0001/kk_1159.root (id bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f) is not in ASO database. Will inject a new transfer and publication request.
Fri, 12 Mar 2021 16:10:06 CET(+0000):INFO:PostJob ASO job description: {'_id': 'bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f',
Fri, 12 Mar 2021 16:40:07 CET(+0000):ERROR:PostJob Stageout job (internal ID bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f) failed with status 'failed'.
Fri, 12 Mar 2021 16:40:07 CET(+0000):INFO:PostJob There were 1 failed/killed transfers: bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f
- bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f:
- bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f:
belforte@vocms059/SPOOL_DIR> grep bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f postjob.1159.1.txt 
Fri, 12 Mar 2021 16:43:20 CET(+0000):INFO:PostJob LFN /store/temp/user/belforte.be1f4dc5be8664cbd145bf008f5399adf42b086f/GenericTTbar/Stefano-Test-210312/210312_145732/0001/kk_1159.root (id bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f) is already in ASO database (file transfer status is 'failed'), but does not correspond to the current job retry. Will inject a new publication request.
 u'id': u'bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f',
 'job_id': u'bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f',
 u'id': u'bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f',
 'job_id': u'bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f',
belforte@vocms059/SPOOL_DIR> 
```

**Comment by belforte** [comment]
this line is fishy
Fri, 12 Mar 2021 16:43:20 CET(+0000):INFO:PostJob LFN /store/temp/user/belforte.be1f4dc5be8664cbd145bf008f5399adf42b086f/GenericTTbar/Stefano-Test-210312/210312_145732/0001/kk_1159.root (id bd578193e5689dd3ebaa8c0b7365a6a5fdf33e449e591a7549dcc10f) is already in ASO database 
(file transfer status is 'failed'), but does not correspond to the current job retry. Will inject a new publication request.

So PJ reset the status to NEW before resubmitting, but when retried the jobs ran at CERN, used local stage out, FTS_transfers was not kicked in and nobody updated the status in the TRANSFERSDB.

Which is what we saw already but dismissed:  #6397

**Comment by belforte** [comment]
Will start by addressing the possible PostJob failure in #6397. Hopefully it will solve this as well.

**Comment by belforte** [comment]
solved via https://github.com/dmwm/CRABServer/pull/6468

**Comment by belforte** [comment]
the fix was put in production with [v3.210318](https://github.com/dmwm/CRABServer/releases/tag/v3.210318) deployed on March 19 as per https://cms-logbook.cern.ch/elog/Analysis+Operations/3437. Since the fix was in the PostJob code it was only effective in tasks submitted after that. But the effect is very clear, since about March 23 the number of files in NEW has stopped to grow. As we drop old DB partitions, the plot should become "clean":
![Screenshot from 2021-03-31 17-44-22](https://user-images.githubusercontent.com/1837785/113173867-1dcad780-924a-11eb-8986-4328062d5fdd.png)
